src/dtls13.c

```diff
@@ -0,0 +1,2442 @@
+            found = 1;
+            _epoch = e->epochNumber;
+        }
+    }
+
+    if (found) {
+        *epoch = _epoch;
+        return 0;
+    }
+
+    return SEQUENCE_ERROR;
+}
+
+/**
+ * Dtls13ParseUnifedRecordLayer() - parse DTLS unified header
+ * @ssl: [in] ssl object
+ * @input: [in] buffer where the header is
+ * @inputSize: [in] size of the input buffer
+ * @hdrInfo: [out] header info struct
+ *
+ * It parse the header and put the relevant information inside @hdrInfo. Further
+ * info: draft43 section 4
+ *
+ * return 0 on success
+ */
+int Dtls13ParseUnifedRecordLayer(WOLFSSL* ssl, const byte *input,
+    word16 inputSize, Dtls13UnifiedHdrInfo *hdrInfo)
+{
+    byte seqLen, hasLength;
+    byte *seqNum;
+    word16 idx;
+    int ret;
+
+    if (input == NULL || inputSize == 0)
+        return BAD_FUNC_ARG;
+
+    if (*input & DTLS13_CID_BIT) {
+        WOLFSSL_MSG("DTLS1.3 header with connection ID. Not supported");
+        return WOLFSSL_NOT_IMPLEMENTED;
+    }
+
+    idx = DTLS13_HDR_FLAGS_SIZE;
+
+    seqLen = (*input & DTLS13_SEQ_LEN_BIT) != 0 ? DTLS13_SEQ_16_LEN
+                                          : DTLS13_SEQ_8_LEN;
+    hasLength = *input & DTLS13_LEN_BIT;
+    hdrInfo->epochBits = *input & EE_MASK;
+
+    idx += seqLen;
+
+    if (inputSize < idx)
+        return BUFFER_ERROR;
+
+    if (hasLength) {
+        if (inputSize < idx + DTLS13_LEN_SIZE)
+            return BUFFER_ERROR;
+
+        ato16(input + idx, &hdrInfo->recordLength);
+        idx += DTLS13_LEN_SIZE;
+
+        /* DTLS message must fit inside a datagram  */
+        if (inputSize < idx + hdrInfo->recordLength)
+            return LENGTH_ERROR;
+    }
+    else {
+        /* length not present. The size of the record is the all the remaining
+           data received with this datagram */
+        hdrInfo->recordLength = inputSize - idx;
+    }
+
+    /* minimum size for a dtls1.3 packet is 16 bytes (to have enough ciphertext
+       to create record number xor mask). (draft 43 - Sec 4.2.3) */
+    if (hdrInfo->recordLength < DTLS13_RN_MASK_SIZE)
+        return LENGTH_ERROR;
+
+    seqNum = (byte*)(input + DTLS13_HDR_FLAGS_SIZE);
+
+    ret = Dtls13EncryptDecryptRecordNumber(
+        ssl, seqNum, seqLen, input + idx, DEPROTECT);
+    if (ret != 0)
+        return ret;
+
+    hdrInfo->headerLength = idx;
+
+    if (seqLen == DTLS13_SEQ_16_LEN) {
+        hdrInfo->seqHiPresent = 1;
+        hdrInfo->seqHi = seqNum[0];
+        hdrInfo->seqLo = seqNum[1];
+    }
+    else {
+        hdrInfo->seqHiPresent = 0;
+        hdrInfo->seqLo = seqNum[0];
+    }
+
+    return 0;
+}
+
+int Dtls13RecordRecvd(WOLFSSL* ssl)
+{
+    int ret;
+
+    if (ssl->curRL.type != handshake)
+        return 0;
+
+    ssl->dtls13FastTimeout = 1;
+    ret = Dtls13RtxAddAck(ssl, ssl->keys.curEpoch64, ssl->keys.curSeq);
+    if (ret != 0)
+        WOLFSSL_MSG("can't save ack fragment");
+
+    return ret;
+}
+
+static void Dtls13RtxMoveToEndOfList(Dtls13RtxRecord **listPtr,
```

Comment:
Thanks now it looks way simpler and cleaner. In the future, If we use list structure in other parts of the code it may be helpful to abstract the data structure and share the code across the codebase.